### ISSUE TITLE
Support storing voting snapshots in the database

### ISSUE THREAD

**Issue by jwasinger** [goerli/parity-goerli#63]
Otherwise, we have to trace back through the chain and rebuild the voting snapshot on startup (which can be very slow depending on where we are in the epoch).

**Comment by jleni** [comment]
I am not sure about the context of what you've been discussing. 
Are you referring to an internal class for Snapshots matching Geth snapshots?

To math the types in the API, I have to create this:
https://github.com/goerli/parity-goerli/pull/58/files#diff-c08ee81f30b2cc1b9b605a93a4267708R47

But having this tracked internally instead would be useful

**Comment by jwasinger** [comment]
I'm referring to the way snapshots are constructed.  Right now, when the client starts up we must traverse the chain backwards to the last epoch in order rebuild the snapshot state:

https://github.com/goerli/parity-goerli/blob/master/ethcore/src/engines/clique/mod.rs#L266-L299

Geth stores voting snapshots on disk to avoid this.  We can do the same.

**Comment by jleni** [comment]
yes, this is a problem for the REST API, because I need to access them by index or hash

**Comment by jleni** [comment]
if this is refactored, it would be amazing if:

- storage allows for fast access via index or hash
- the data structures match Geth.

Otherwise, searching/wrapping/converting on each REST call will be inefficient. 
We can the types/tests from that file I linked earlier.

**Comment by jwasinger** [comment]
The voting snapshot is only restored during client startup so there should be no overhead for the RPC API AFAIK

**Comment by jwasinger** [comment]
I haven't had time to review the RPC API changes yet so I could be wrong.

**Comment by jleni** [comment]
Yes, I have just started. I will be working on this in the next few days. It is much better now that things got merged and are a bit more stable.

I assume then that snapshots are accessible here:
https://github.com/goerli/parity-goerli/blob/941d454fd616badd76e76c6dc7e3e8d1fcd0e914/ethcore/src/engines/clique/mod.rs#L160

I may need a way to get them by index. I need to review the code a bit more.

**Comment by jwasinger** [comment]
Cool.  Looking forward to seeing what you come up with :)

**Comment by jwasinger** [comment]
My inclination is that this issue is nice to have but the implementation should be merged separately from the main clique PR in order to avoid more review overhead from the Parity side.

**Comment by jleni** [comment]
would it be possible to make this public?

https://github.com/goerli/parity-goerli/blob/5fa0228154b223694622b48f7d3cc73046c00bf3/ethcore/src/engines/clique/mod.rs#L254

**Comment by jwasinger** [comment]
Yes.  I think so.

**Comment by jwasinger** [comment]
Hey @jleni looks like this might be easy to add via the "block metadata feature" (see the related comment on the PR above).  I'm going to look into this.
